core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -2219,11 +2219,46 @@ object ZIO {
    *
    * For a sequential version of this method, see `foreach`.
    */
-  def foreachPar[R, E, A, B](as: Iterable[A])(fn: A => ZIO[R, E, B]): ZIO[R, E, List[B]] =
-    as.foldRight[ZIO[R, E, List[B]]](effectTotal(Nil)) { (a, io) =>
-        fn(a).zipWithPar(io)((b, bs) => b :: bs)
-      }
-      .refailWithTrace
+  final def foreachPar[R, E, A, B](as: Iterable[A])(fn: A => ZIO[R, E, B]): ZIO[R, E, List[B]] = {
+    val n         = as.size
+    val resultArr = new Array[Any](n)
+    val resultList: ZIO[R, E, List[B]] = for {
+      parentId <- ZIO.fiberId
+      latch    <- CountdownLatch.make(n)
+      failed   <- Promise.make[Nothing, Unit]
+      cause    <- Ref.make[Cause[E]](Cause.empty)
+      fibers <- as
+                 .foldLeft[(Int, URIO[R, List[Fiber[E, _]]])]((0, URIO.succeed(Nil))) {
+                   case ((i, acc), a) =>
+                     val task = ZIOFn(fn)(
+                       (x: A) =>
+                         fn(x).traced
+                           .foldCauseM(
+                             c => cause.update(cs => cs && c) *> failed.succeed(()).unit,
+                             b => ZIO.effectTotal(resultArr(i) = b)
```

Comment:
Could we implement `foreachPar` in terms of `foreachPar_`? The same way as `mergeAllPar` is implemented in #2614, but putting results into Array instead of Ref.

Reply:
Yes, that would be entirely reasonable. I'd be more than happy to give up on this PR if you'd like to do it on top of your version.

Reply:
I'm not sure if I will be able to do it soon :(
I guess, you could do it yourself in this PR too :)